L1Trigger/TrackFindingTMTT/src/Histos.cc

```diff
@@ -0,0 +1,1776 @@
+    hisNumEtaSecsPerStub_ =
+        inputDir.make<TH1F>("NumEtaSecPerStub", "; Number of #eta sectors each stub appears in", 20, -0.5, 19.5);
+    hisNumPhiSecsPerStub_ =
+        inputDir.make<TH1F>("NumPhiSecPerStub", "; Number of #phi sectors each stub appears in", 20, -0.5, 19.5);
+
+    // Count stubs per (eta,phi) sector.
+    hisNumStubsPerSec_ = inputDir.make<TH1F>("NumStubsPerSec", "; Number of stubs per sector", 250, -0.5, 249.5);
+
+    return inputDir;
+  }
+
+  //=== Fill histograms checking if (eta,phi) sector definition choices are good.
+
+  void Histos::fillEtaPhiSectors(const InputData& inputData, const matrix<unique_ptr<Sector>>& mSectors) {
+    // Allow only one thread to run this function at a time
+    std::lock_guard<std::mutex> myGuard(myMutex[1]);
+
+    const list<const Stub*>& vStubs = inputData.stubsConst();
+    //const list<TP>& vTPs = inputData.getTPs();
+
+    //=== Loop over all stubs, counting how many sectors each one appears in.
+
+    for (const Stub* stub : vStubs) {
+      // Number of (eta,phi), phi & eta sectors containing this stub.
+      unsigned int nEtaSecs = 0;
+      unsigned int nPhiSecs = 0;
+
+      // Loop over (eta, phi) sectors.
+      for (unsigned int iPhiSec = 0; iPhiSec < numPhiSectors_; iPhiSec++) {
+        for (unsigned int iEtaReg = 0; iEtaReg < numEtaRegions_; iEtaReg++) {
+          const Sector* sector = mSectors(iPhiSec, iEtaReg).get();
+
+          // Check if sector contains stub stub, and if so count it.
+          // Take care to just use one eta (phi) typical region when counting phi (eta) sectors.
+          if (iPhiSec == 0 && sector->insideEta(stub))
+            nEtaSecs++;
+          if (iEtaReg == 0 && sector->insidePhi(stub))
+            nPhiSecs++;
+        }
+      }
+
+      // Plot number of sectors each stub appears in.
+      hisNumEtaSecsPerStub_->Fill(nEtaSecs);
+      hisNumPhiSecsPerStub_->Fill(nPhiSecs);
+    }
+
+    //=== Loop over all sectors, counting the stubs in each one.
+    for (unsigned int iEtaReg = 0; iEtaReg < numEtaRegions_; iEtaReg++) {
+      for (unsigned int iPhiSec = 0; iPhiSec < numPhiSectors_; iPhiSec++) {
+        const Sector* sector = mSectors(iPhiSec, iEtaReg).get();
+
+        unsigned int nStubs = 0;
+        for (const Stub* stub : vStubs) {
+          if (sector->inside(stub))
+            nStubs++;
+        }
+        hisNumStubsPerSec_->Fill(nStubs);
+      }
+    }
+  }
+
+  //=== Book histograms checking filling of r-phi HT array.
+
+  TFileDirectory Histos::bookRphiHT() {
+    TFileDirectory inputDir = fs_->mkdir("HTrphi");
+
+    return inputDir;
+  }
+
+  //=== Fill histograms checking filling of r-phi HT array.
+
+  void Histos::fillRphiHT(const matrix<unique_ptr<HTrphi>>& mHtRphis) {
+    //--- Loop over (eta,phi) sectors, counting the number of stubs in the HT array of each.
+
+    // Allow only one thread to run this function at a time
+    std::lock_guard<std::mutex> myGuard(myMutex[2]);
```

Comment:
Having a `lock_guard` at the end of the function doesn't make sense - this should just be removed.

Reply:
The empty functions named fillxyz() in Histos are provided so that the user can add, if they wish, histograms for studying particular elements of the tracking chain. The presence of the mutex in these functions is useful to ensure that if the user does this, their code will be thread safe. I've commented out the mutex however.